### ISSUE TITLE
Xgboost results different depending on column order

### ISSUE THREAD

**Issue by vSkullv** [dmlc/xgboost#2924]
Hi,

I was curious why I get different results depend on the order of the columns in the design matrix.
Here is an example:
1) Fit a model
2) Change the order of the columns
3) Fit a model
4) The resulting 2 models are different.

library("xgboost")

data(iris)

iris_y = as.integer(iris$Sepal.Width)

#Model 1##########################################################################
set.seed(0)
iris_X = data.frame(model.matrix(Sepal.Width~.-1,data=iris))

iris.CountPoisson1 = xgboost(data = as.matrix(iris_X), 
                             label = iris_y,
                             objective="count:poisson",
                             nrounds = 10,
                             learning_rate = 0.3)
importance1 = xgb.importance(iris.CountPoisson1, feature_names = colnames(iris_X));

#Model 2##########################################################################

set.seed(0)
iris_X2 = iris_X[, c(6,4,2,3,1,5)]

stopifnot(sum(iris_X)==sum(iris_X2),"Error: Totals must be equal.")
iris.CountPoisson2 = xgboost(data = as.matrix(iris_X2), 
                             label = iris_y, 
                             objective="count:poisson",
                             nrounds = 10,
                             learning_rate = 0.3)

importance2 = xgb.importance(iris.CountPoisson2, feature_names = colnames(iris_X2));

importance1
importance2

Can you let me know why this is the case?

Thanks,
Hasan

**Comment by khotilov** [comment]
In the iris dataset, each, Petal.Length and Petal.Width, is able to perfectly separate the setosa class:
```
lattice::splom(iris, col=iris$Species)
```
And when regressing over them to determine the Sepal.Width, all those three features would result in the same gain when used for the first split in a tree. When xgboost sees multiple features with the same best gain, the first one in the column order would be used is used.

Also, for the Poissoin objective it is especially useful to set the base_score=mean(iris_y) to improve the convergence.
